modules/dhcp/providers/server/isc.rb

```diff
@@ -104,38 +91,72 @@ def loadSubnetData subnet
             opts.merge! parse_record_options(data)
           end
           next if opts[:mac].nil?
-          # delete expired (explicitly) expired leases
-          if opts[:state] == "free" || (opts[:next_state] == "free" && opts[:ends] && opts[:ends] < Time.now)
-            record = find_record_by_ip(subnet, ip, 'lease')
-            subnet.delete record if record
+
+          subnet = @service.find_subnet(ip)
+          next unless subnet
+          ret_val << Proxy::DHCP::Lease.new(opts.merge(:subnet => subnet, :ip => ip))
+        end
+      end
+
+      ret_val
+    end
+
+    def load_subnet_data(records)
+      records.each do |record|
+
+        case record
+        when Proxy::DHCP::Reservation
+          if record.options[:deleted]
+            record = @service.find_host_by_hostname(record.subnet_address, record.name)
+            @service.delete_host(record) if record
             next
           end
-          if subnet.include? ip
-            # delete possible duplicities 
-            if dupe = subnet.has_mac?(opts[:mac], :lease)
-              subnet.delete(dupe)
-            end
-            if dupe = subnet.has_ip?(ip, :lease)
-              subnet.delete(dupe)
-            end
-            # and add the record
-            Proxy::DHCP::Lease.new(opts.merge(:subnet => subnet, :ip => ip))
+          if dupe = @service.find_host_by_mac(record.subnet_address, record.mac)
+            @service.delete_host(dupe)
           end
+          if dupe = @service.find_host_by_ip(record.subnet_address, record.ip)
+            @service.delete_host(dupe)
+          end
+
+          @service.add_host(record.subnet_address, record)
+        when Proxy::DHCP::Lease
+          if record.options[:state] == "free" || (record.options[:next_state] == "free" && record.options[:ends] && record.options[:ends] < Time.now)
+            record = @service.find_lease_by_ip(record.subnet_address, record.ip)
+            @service.delete_lease(record) if record
+            next
+          end
+
+          if dupe = @service.find_lease_by_mac(record.subnet_address, record.mac)
+            @service.delete_lease(dupe)
+          end
+          if dupe = @service.find_lease_by_ip(record.subnet_address, record.ip)
+            @service.delete_lease(dupe)
+          end
+
+          @service.add_lease(record.subnet_address, record)
         end
       end
-      report "Enumerated hosts on #{subnet.network}"
     end
 
-    private
+    def loadSubnetData subnet
+      load_subnet_data(parse_config_and_leases)
+    end
+
+    def load_subnets
```

Comment:
This is incredibly similar naming to loadSubnets, can they be substantially different please?  Or incorporate this straight into loadSubnets, or rename loadSubnets to load_subnets.  (It would be nice to remove the camelCase names.)